Fix python3 traceback in netconf connection plugin

abadger opened the pull request:
The ncclient API that's being used here requires request is a text
string on Python3 and appears to anticipate a byte string on python2.

This was reported in a comment on #24431:

https://github.com/ansible/ansible/pull/24431#issuecomment-300999221

##### ISSUE TYPE

 - Bugfix Pull Request


##### COMPONENT NAME
lib/ansible/plugins/connection/netconf.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
devel 2.3
```


##### ADDITIONAL INFORMATION
From the report in https://github.com/ansible/ansible/pull/24431#issuecomment-300999221

> I'll see about setting up some docker containers for testing both on the special case of Arch with Python3 as well as the more common scenario with ansible_python_interpreter = /usr/bin/python3 on Ubuntu.

> With the latest commits:
```
2017-05-12 08:46:03,521 p=6697 u=tjuberg |  Using /home/tjuberg/work/gitlab/bachelor_ansible/ansible.cfg as config file 
2017-05-12 08:46:03,722 p=6697 u=tjuberg |  Loading callback plugin default of type stdout, v2.0 from /home/tjuberg/.local/badger-ansible/lib/ansible/plugins/callback/__init__.py 
2017-05-12 08:46:03,777 p=6697 u=tjuberg |  PLAYBOOK: junos_facts.yml ****************************************************************************************************** 
2017-05-12 08:46:03,777 p=6697 u=tjuberg |  1 plays in junos_facts.yml 
2017-05-12 08:46:03,811 p=6697 u=tjuberg |  PLAY [Get JunOS facts] ********************************************************************************************************* 
2017-05-12 08:46:03,825 p=6697 u=tjuberg |  META: ran handlers 
2017-05-12 08:46:03,827 p=6697 u=tjuberg |  TASK [Gather Junos facts] ****************************************************************************************************** 
2017-05-12 08:46:03,827 p=6697 u=tjuberg |  task path: /home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.yml:9 
2017-05-12 08:46:04,428 p=6706 u=tjuberg |  creating new control socket for host xxx.xxx.xxx.xxx:22 as user ansible 
2017-05-12 08:46:04,428 p=6706 u=tjuberg |  control socket path is /home/tjuberg/.ansible/pc/d930766c35 
2017-05-12 08:46:04,428 p=6706 u=tjuberg |  current working directory is /home/tjuberg/work/gitlab/bachelor_ansible 
2017-05-12 08:46:04,429 p=6706 u=tjuberg |  using connection plugin netconf 
2017-05-12 08:46:04,530 p=6706 u=tjuberg |  network_os is set to junos 
2017-05-12 08:46:04,530 p=6706 u=tjuberg |  ssh connection done, stating ncclient 
2017-05-12 08:46:04,544 ncclient.transport.ssh Connected (version 2.0, client OpenSSH_6.6.1) 
2017-05-12 08:46:04,866 ncclient.transport.ssh Authentication (password) successful! 
2017-05-12 08:46:05,019 ncclient.transport.session initialized: session-id=19666 | server_capabilities=<dict_keyiterator object at 0x7f348c13ef48> 
2017-05-12 08:46:05,019 p=6706 u=tjuberg |  ncclient manager object created successfully 
2017-05-12 08:46:05,020 p=6706 u=tjuberg |  connection established to xxx.xxx.xxx.xxx in 0:00:00.590981 
2017-05-12 08:46:05,427 p=6706 u=tjuberg |  incoming request accepted on persistent socket 
2017-05-12 08:46:05,427 p=6706 u=tjuberg |  socket operation is CONTEXT 
2017-05-12 08:46:05,428 p=6706 u=tjuberg |  socket operation is EXEC 
2017-05-12 08:46:05,429 p=6706 u=tjuberg |  socket operation completed with rc 255 
2017-05-12 08:46:05,493 p=6697 u=tjuberg |  open_session() returned 255 b'' b'Traceback (most recent call last):\n  
File "/home/tjuberg/.local/badger-ansible/bin/ansible-connection", line 199, in run\n
(rc, stdout, stderr) = self.conn.exec_command(cmd)\n
File "/home/tjuberg/.local/badger-ansible/lib/ansible/plugins/connection/__init__.py", line 52, in wrapped\n
return func(self, *args, **kwargs)\n
File "/home/tjuberg/.local/badger-ansible/lib/ansible/plugins/connection/netconf.py", line 113, in exec_command\n
req = to_ele(request)\n
File "/home/tjuberg/.local/lib/python3.6/site-packages/ncclient/xml_.py", line 109, in to_ele\n
return x if etree.iselement(x) else etree.fromstring(x.encode(\'UTF-8\'), parser=parser)\n
AttributeError: \'bytes\' object has no attribute \'encode\'\n' 
2017-05-12 08:46:05,495 p=6697 u=tjuberg |  fatal: [trd-gw]: FAILED! => { 
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell", 
    "rc": 255 
} 
2017-05-12 08:46:05,496 p=6697 u=tjuberg |  	to retry, use: --limit @/home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.retry 

2017-05-12 08:46:05,497 p=6697 u=tjuberg |  PLAY RECAP ********************************************************************************************************************* 
2017-05-12 08:46:05,497 p=6697 u=tjuberg |  trd-gw                     : ok=0    changed=0    unreachable=0    failed=1    
2017-05-12 08:46:35,429 p=6706 u=tjuberg |  shutting down control socket, connection was active for 0:00:31.000909 secs 
2017-05-12 08:46:35,430 ncclient.operations.rpc Requesting 'CloseSession' 
```

abadger committed:
Fix python3 traceback in netconf connection plugin

The ncclient API that's being used here requires request is a text
string on Python3 and appears to anticipate a byte string on python2.

This was reported in a comment on #24431:

https://github.com/ansible/ansible/pull/24431#issuecomment-300999221

abadger commented:
@TJuberg Here's the PR I was promising you.  It should resolve the traceback you reported but there could be plenty more python3 bugs to resolve in this code.  Let me know the results of your testing and I'll update the PR with more fixes.

TJuberg commented:
Full log using -vvvv on your py3-netconf-connection branch
```
2017-05-13 17:11:45,081 p=15074 u=tjuberg |  Using /home/tjuberg/work/gitlab/bachelor_ansible/ansible.cfg as config file
2017-05-13 17:11:45,415 p=15074 u=tjuberg |  Loading callback plugin default of type stdout, v2.0 from /home/tjuberg/.local/badger-ansible/lib/ansible/plugins/callback/__init__.py
2017-05-13 17:11:45,516 p=15074 u=tjuberg |  PLAYBOOK: junos_facts.yml *****************************************************************************************************************************
2017-05-13 17:11:45,516 p=15074 u=tjuberg |  1 plays in junos_facts.yml
2017-05-13 17:11:45,556 p=15074 u=tjuberg |  PLAY [Get JunOS facts] ********************************************************************************************************************************
2017-05-13 17:11:45,576 p=15074 u=tjuberg |  META: ran handlers
2017-05-13 17:11:45,577 p=15074 u=tjuberg |  TASK [Gather Junos facts] *****************************************************************************************************************************
2017-05-13 17:11:45,577 p=15074 u=tjuberg |  task path: /home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.yml:9
2017-05-13 17:11:45,627 p=15074 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/junos.py
2017-05-13 17:11:45,631 p=15074 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/basic.py
2017-05-13 17:11:45,631 p=15074 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/netconf.py
2017-05-13 17:11:45,632 p=15074 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/six/__init__.py
2017-05-13 17:11:45,642 p=15074 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/_text.py
2017-05-13 17:11:45,708 p=15074 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/pycompat24.py
2017-05-13 17:11:45,713 p=15074 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/connection.py
2017-05-13 17:11:45,718 p=15074 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/six/_six.py
2017-05-13 17:11:45,745 p=15074 u=tjuberg |  Using module file /home/tjuberg/.local/badger-ansible/lib/ansible/modules/network/junos/junos_facts.py
2017-05-13 17:11:46,096 p=15074 u=tjuberg |  The full traceback is:
Traceback (most recent call last):
  File "/tmp/tjuberg/ansible_gr_7v30b/ansible_module_junos_facts.py", line 280, in <module>
    main()
  File "/tmp/tjuberg/ansible_gr_7v30b/ansible_module_junos_facts.py", line 268, in main
    inst.populate()
  File "/tmp/tjuberg/ansible_gr_7v30b/ansible_module_junos_facts.py", line 117, in populate
    reply = self.rpc('get-software-information')
  File "/tmp/tjuberg/ansible_gr_7v30b/ansible_module_junos_facts.py", line 105, in rpc
    return send_request(self.module, Element(rpc))
  File "/tmp/tjuberg/ansible_gr_7v30b/ansible_modlib.zip/ansible/module_utils/netconf.py", line 41, in send_request
  File "/usr/lib/python3.6/xml/etree/ElementTree.py", line 1314, in XML
    parser.feed(text)
xml.etree.ElementTree.ParseError: syntax error: line 1, column 0

2017-05-13 17:11:46,097 p=15074 u=tjuberg |  fatal: [mx104]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/tjuberg/ansible_gr_7v30b/ansible_module_junos_facts.py\", line 280, in <module>\n    main()\n  File \"/tmp/tjuberg/ansible_gr_7v30b/ansible_module_junos_facts.py\", line 268, in main\n    inst.populate()\n  File \"/tmp/tjuberg/ansible_gr_7v30b/ansible_module_junos_facts.py\", line 117, in populate\n    reply = self.rpc('get-software-information')\n  File \"/tmp/tjuberg/ansible_gr_7v30b/ansible_module_junos_facts.py\", line 105, in rpc\n    return send_request(self.module, Element(rpc))\n  File \"/tmp/tjuberg/ansible_gr_7v30b/ansible_modlib.zip/ansible/module_utils/netconf.py\", line 41, in send_request\n  File \"/usr/lib/python3.6/xml/etree/ElementTree.py\", line 1314, in XML\n    parser.feed(text)\nxml.etree.ElementTree.ParseError: syntax error: line 1, column 0\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}
2017-05-13 17:11:46,098 p=15074 u=tjuberg |  	to retry, use: --limit @/home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.retry

2017-05-13 17:11:46,098 p=15074 u=tjuberg |  PLAY RECAP ********************************************************************************************************************************************
```

abadger committed:
Fix potential corruption of the returned data if non-utf8 is returned.

It's the caller's responsibility to decide how to deal with the malformed data.

abadger commented:
This error is odd.  It looks like the remote end is returning bad data.  My latest change will fix a problem if the data contains non-utf8 (but this will only fix python3, python2 should have been failing in this case [and if it did, it will continue to do so....) I'm not too hopeful about this fixing it, though, because the ElementTree error indicates that the malformed data is right at the very beginning (line 1, column 0).... Just depends on whether ElementTree starts counting lines with 1 or 0.

If the traceback is the same (or in the same area of the code) then I'll push some instrumented code so we can look at the data value that's causing the error.

TJuberg commented:
Updated trace.

For reference, on the router end the logs indicate that the actual request never gets sent and that the crash appears to happen the moment the netconf connection is established. Running this branch on my python2 test system does not produce any errors for the same play and equipment.

```
2017-05-15 09:24:19,699 p=7579 u=tjuberg |  Using /home/tjuberg/work/gitlab/bachelor_ansible/ansible.cfg as config file
2017-05-15 09:24:19,952 p=7579 u=tjuberg |  Loading callback plugin default of type stdout, v2.0 from /home/tjuberg/.local/badger-ansible/lib/ansible/plugins/callback/__init__.py
2017-05-15 09:24:20,013 p=7579 u=tjuberg |  PLAYBOOK: junos_facts.yml ****************************************************************************
2017-05-15 09:24:20,013 p=7579 u=tjuberg |  1 plays in junos_facts.yml
2017-05-15 09:24:20,051 p=7579 u=tjuberg |  PLAY [Get JunOS facts] *******************************************************************************
2017-05-15 09:24:20,064 p=7579 u=tjuberg |  META: ran handlers
2017-05-15 09:24:20,066 p=7579 u=tjuberg |  TASK [Gather Junos facts] ****************************************************************************
2017-05-15 09:24:20,066 p=7579 u=tjuberg |  task path: /home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.yml:9
2017-05-15 09:24:20,648 p=7588 u=tjuberg |  creating new control socket for host xxx.xxx.xxx.xxx:22 as user ansible
2017-05-15 09:24:20,648 p=7588 u=tjuberg |  control socket path is /home/tjuberg/.ansible/pc/b8a6122e69
2017-05-15 09:24:20,648 p=7588 u=tjuberg |  current working directory is /home/tjuberg/work/gitlab/bachelor_ansible
2017-05-15 09:24:20,648 p=7588 u=tjuberg |  using connection plugin netconf
2017-05-15 09:24:20,739 p=7588 u=tjuberg |  network_os is set to junos
2017-05-15 09:24:20,739 p=7588 u=tjuberg |  ssh connection done, stating ncclient
2017-05-15 09:24:20,792 ncclient.transport.ssh Connected (version 2.0, client OpenSSH_7.1)
2017-05-15 09:24:21,927 ncclient.transport.ssh Authentication (password) successful!
2017-05-15 09:24:22,755 ncclient.transport.session initialized: session-id=75163 | server_capabilities=<dict_keyiterator object at 0x7ff25486a318>
2017-05-15 09:24:22,755 p=7588 u=tjuberg |  ncclient manager object created successfully
2017-05-15 09:24:22,755 p=7588 u=tjuberg |  connection established to xxx.xxx.xxx.xxx in 0:00:02.107191
2017-05-15 09:24:23,649 p=7588 u=tjuberg |  incoming request accepted on persistent socket
2017-05-15 09:24:23,649 p=7588 u=tjuberg |  socket operation is CONTEXT
2017-05-15 09:24:23,650 p=7588 u=tjuberg |  socket operation is EXEC
2017-05-15 09:24:23,650 p=7588 u=tjuberg |  socket operation completed with rc 0
2017-05-15 09:24:23,718 p=7579 u=tjuberg |  open_session() returned 0 b'ok' b''
2017-05-15 09:24:23,726 p=7579 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/netconf.py
2017-05-15 09:24:23,726 p=7579 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/six/__init__.py
2017-05-15 09:24:23,727 p=7579 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/junos.py
2017-05-15 09:24:23,730 p=7579 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/basic.py
2017-05-15 09:24:23,732 p=7579 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/connection.py
2017-05-15 09:24:23,734 p=7579 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/_text.py
2017-05-15 09:24:23,738 p=7579 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/six/_six.py
2017-05-15 09:24:23,809 p=7579 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/pycompat24.py
2017-05-15 09:24:23,812 p=7579 u=tjuberg |  Using module file /home/tjuberg/.local/badger-ansible/lib/ansible/modules/network/junos/junos_facts.py
2017-05-15 09:24:24,028 p=7588 u=tjuberg |  incoming request accepted on persistent socket
2017-05-15 09:24:24,028 p=7588 u=tjuberg |  socket operation is EXEC
2017-05-15 09:24:24,028 p=7588 u=tjuberg |  socket operation completed with rc 1
2017-05-15 09:24:24,049 p=7579 u=tjuberg |  The full traceback is:
Traceback (most recent call last):
  File "/tmp/tjuberg/ansible__pece251/ansible_module_junos_facts.py", line 280, in <module>
    main()
  File "/tmp/tjuberg/ansible__pece251/ansible_module_junos_facts.py", line 268, in main
    inst.populate()
  File "/tmp/tjuberg/ansible__pece251/ansible_module_junos_facts.py", line 117, in populate
    reply = self.rpc('get-software-information')
  File "/tmp/tjuberg/ansible__pece251/ansible_module_junos_facts.py", line 105, in rpc
    return send_request(self.module, Element(rpc))
  File "/tmp/tjuberg/ansible__pece251/ansible_modlib.zip/ansible/module_utils/netconf.py", line 41, in send_request
  File "/usr/lib/python3.6/xml/etree/ElementTree.py", line 1314, in XML
    parser.feed(text)
xml.etree.ElementTree.ParseError: syntax error: line 1, column 0

2017-05-15 09:24:24,050 p=7579 u=tjuberg |  fatal: [mx104]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/tjuberg/ansible__pece251/ansible_module_junos_facts.py\", line 280, in <module>\n    main()\n  File \"/tmp/tjuberg/ansible__pece251/ansible_module_junos_facts.py\", line 268, in main\n    inst.populate()\n  File \"/tmp/tjuberg/ansible__pece251/ansible_module_junos_facts.py\", line 117, in populate\n    reply = self.rpc('get-software-information')\n  File \"/tmp/tjuberg/ansible__pece251/ansible_module_junos_facts.py\", line 105, in rpc\n    return send_request(self.module, Element(rpc))\n  File \"/tmp/tjuberg/ansible__pece251/ansible_modlib.zip/ansible/module_utils/netconf.py\", line 41, in send_request\n  File \"/usr/lib/python3.6/xml/etree/ElementTree.py\", line 1314, in XML\n    parser.feed(text)\nxml.etree.ElementTree.ParseError: syntax error: line 1, column 0\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}
2017-05-15 09:24:24,051 p=7579 u=tjuberg |  	to retry, use: --limit @/home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.retry

2017-05-15 09:24:24,051 p=7579 u=tjuberg |  PLAY RECAP *******************************************************************************************
2017-05-15 09:24:24,051 p=7579 u=tjuberg |  mx104                      : ok=0    changed=0    unreachable=0    failed=1   
2017-05-15 09:24:54,029 p=7588 u=tjuberg |  shutting down control socket, connection was active for 0:00:33.380660 secs
2017-05-15 09:24:54,030 ncclient.operations.rpc Requesting 'CloseSession'

```

abadger committed:
Instrument code to get some better info

abadger committed:
Add traceback to returned info

abadger commented:
Okay, I pushed a small change to give us more information about the failure.  Run it and let's see what the data is.

TJuberg commented:
Note the typo on line 33 in the import statement btw.

Updated log

```
2017-05-15 15:37:01,133 p=11225 u=tjuberg |  Using /home/tjuberg/work/gitlab/bachelor_ansible/ansible.cfg as config file
2017-05-15 15:37:01,993 p=11225 u=tjuberg |  Loading callback plugin default of type stdout, v2.0 from /home/tjuberg/.local/badger-ansible/lib/ansible/plugins/callback/__init__.py
2017-05-15 15:37:02,161 p=11225 u=tjuberg |  PLAYBOOK: junos_facts.yml ****************************************************************************
2017-05-15 15:37:02,161 p=11225 u=tjuberg |  1 plays in junos_facts.yml
2017-05-15 15:37:02,261 p=11225 u=tjuberg |  PLAY [Get JunOS facts] *******************************************************************************
2017-05-15 15:37:02,289 p=11225 u=tjuberg |  META: ran handlers
2017-05-15 15:37:02,296 p=11225 u=tjuberg |  TASK [Gather Junos facts] ****************************************************************************
2017-05-15 15:37:02,297 p=11225 u=tjuberg |  task path: /home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.yml:9
2017-05-15 15:37:04,552 p=11234 u=tjuberg |  creating new control socket for host xxx.xxx.xxx.xxx:22 as user ansible
2017-05-15 15:37:04,558 p=11234 u=tjuberg |  control socket path is /home/tjuberg/.ansible/pc/b8a6122e69
2017-05-15 15:37:04,558 p=11234 u=tjuberg |  current working directory is /home/tjuberg/work/gitlab/bachelor_ansible
2017-05-15 15:37:04,558 p=11234 u=tjuberg |  using connection plugin netconf
2017-05-15 15:37:04,932 p=11234 u=tjuberg |  network_os is set to junos
2017-05-15 15:37:04,933 p=11234 u=tjuberg |  ssh connection done, stating ncclient
2017-05-15 15:37:04,985 ncclient.transport.ssh Connected (version 2.0, client OpenSSH_7.1)
2017-05-15 15:37:06,436 ncclient.transport.ssh Authentication (password) successful!
2017-05-15 15:37:07,264 ncclient.transport.session initialized: session-id=79556 | server_capabilities=<dict_keyiterator object at 0x7f5a43268b38>
2017-05-15 15:37:07,265 p=11234 u=tjuberg |  ncclient manager object created successfully
2017-05-15 15:37:07,266 p=11234 u=tjuberg |  connection established to xxx.xxx.xxx.xxx in 0:00:02.707578
2017-05-15 15:37:07,536 p=11234 u=tjuberg |  incoming request accepted on persistent socket
2017-05-15 15:37:07,537 p=11234 u=tjuberg |  socket operation is CONTEXT
2017-05-15 15:37:07,538 p=11234 u=tjuberg |  socket operation is EXEC
2017-05-15 15:37:07,538 p=11234 u=tjuberg |  socket operation completed with rc 0
2017-05-15 15:37:07,673 p=11225 u=tjuberg |  open_session() returned 0 b'ok' b''
2017-05-15 15:37:07,689 p=11225 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/junos.py
2017-05-15 15:37:07,690 p=11225 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/netconf.py
2017-05-15 15:37:07,697 p=11225 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/basic.py
2017-05-15 15:37:07,698 p=11225 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/six/__init__.py
2017-05-15 15:37:07,716 p=11225 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/_text.py
2017-05-15 15:37:07,725 p=11225 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/connection.py
2017-05-15 15:37:07,895 p=11225 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/pycompat24.py
2017-05-15 15:37:07,906 p=11225 u=tjuberg |  Using module_utils file /home/tjuberg/.local/badger-ansible/lib/ansible/module_utils/six/_six.py
2017-05-15 15:37:07,950 p=11225 u=tjuberg |  Using module file /home/tjuberg/.local/badger-ansible/lib/ansible/modules/network/junos/junos_facts.py
2017-05-15 15:37:08,684 p=11234 u=tjuberg |  incoming request accepted on persistent socket
2017-05-15 15:37:08,685 p=11234 u=tjuberg |  socket operation is EXEC
2017-05-15 15:37:08,685 p=11234 u=tjuberg |  socket operation completed with rc 1
2017-05-15 15:37:08,727 p=11225 u=tjuberg |  The full traceback is:
Traceback (most recent call last):
  File "/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/netconf.py", line 46, in send_request
  File "/usr/lib/python3.6/xml/etree/ElementTree.py", line 1314, in XML
    parser.feed(text)
xml.etree.ElementTree.ParseError: syntax error: line 1, column 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/tjuberg/ansible_zqwttysq/ansible_module_junos_facts.py", line 280, in <module>
    main()
  File "/tmp/tjuberg/ansible_zqwttysq/ansible_module_junos_facts.py", line 268, in main
    inst.populate()
  File "/tmp/tjuberg/ansible_zqwttysq/ansible_module_junos_facts.py", line 180, in populate
    reply = send_request(self.module, ele)
  File "/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/netconf.py", line 48, in send_request
  File "/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/basic.py", line 2027, in fail_json
  File "/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/basic.py", line 2007, in _return_formatted
  File "/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/basic.py", line 413, in remove_values
  File "/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/basic.py", line 413, in <genexpr>
  File "/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/basic.py", line 424, in remove_values
TypeError: Value of unknown type: <class 'function'>, <function format_exc at 0x7fb5b24bec80>

2017-05-15 15:37:08,728 p=11225 u=tjuberg |  fatal: [mx104]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/netconf.py\", line 46, in send_request\n  File \"/usr/lib/python3.6/xml/etree/ElementTree.py\", line 1314, in XML\n    parser.feed(text)\nxml.etree.ElementTree.ParseError: syntax error: line 1, column 0\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/tmp/tjuberg/ansible_zqwttysq/ansible_module_junos_facts.py\", line 280, in <module>\n    main()\n  File \"/tmp/tjuberg/ansible_zqwttysq/ansible_module_junos_facts.py\", line 268, in main\n    inst.populate()\n  File \"/tmp/tjuberg/ansible_zqwttysq/ansible_module_junos_facts.py\", line 180, in populate\n    reply = send_request(self.module, ele)\n  File \"/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/netconf.py\", line 48, in send_request\n  File \"/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/basic.py\", line 2027, in fail_json\n  File \"/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/basic.py\", line 2007, in _return_formatted\n  File \"/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/basic.py\", line 413, in remove_values\n  File \"/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/basic.py\", line 413, in <genexpr>\n  File \"/tmp/tjuberg/ansible_zqwttysq/ansible_modlib.zip/ansible/module_utils/basic.py\", line 424, in remove_values\nTypeError: Value of unknown type: <class 'function'>, <function format_exc at 0x7fb5b24bec80>\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}
2017-05-15 15:37:08,733 p=11225 u=tjuberg |  	to retry, use: --limit @/home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.retry

2017-05-15 15:37:08,734 p=11225 u=tjuberg |  PLAY RECAP *******************************************************************************************
2017-05-15 15:37:08,734 p=11225 u=tjuberg |  mx104                      : ok=0    changed=0    unreachable=0    failed=1   
```

abadger committed:
fix import typo in instrumentation patch

abadger commented:
Thanks for pointing out the typo.   Looks like there's too much information to use our standard fail_json() here.  This code is running on localhost?  Maybe I can change it to output to a file in /tmp/ and then you can upload those somewhere?

abadger committed:
Alternate means of instrumenting code

abadger committed:
Remember to_bytes import

TJuberg commented:
Running on localhost yes. Would be no problem writing to file under either /tmp or out/  won't be able to test until I get back in the office tomorrow however.

mattclay commented:
CI failure in python 3.x unit tests:

```
>       self.assertEqual('<test/>', out)
E       AssertionError: '<test/>' != b'<test/>'

test/units/plugins/connection/test_netconf.py:103: AssertionError

>       self.assertEqual('', out)
E       AssertionError: '' != b''

test/units/plugins/connection/test_netconf.py:118: AssertionError
```

TJuberg commented:
The files unfortunately contains no additional information, traceback remains the same
```
  File "/tmp/tjuberg/ansible_pk7icw6q/ansible_modlib.zip/ansible/module_utils/netconf.py", line 46, in send_request
    error_root = fromstring(err)
  File "/usr/lib/python3.6/xml/etree/ElementTree.py", line 1314, in XML
    parser.feed(text)
  File "<string>", line None
xml.etree.ElementTree.ParseError: syntax error: line 1, column 0
```
I suspect this could be related to the actual error occurring within the forked ansible-connection process.

I am attaching a redacted strace file from the process which should contain everything if you know where to look.

abadger commented:
Hey, I'm heading to pycon today.  Will be on and offline for 10 days.  @ganeshrn are you able to take over troubleshooting this one?

ganeshrn commented:
@abadger Ack, will look into this issue.

ganeshrn commented:
@TJuberg Can you please test with this patch https://gist.github.com/ganeshrn/9848f9db3b965f2dbfd61b8080408305

I have added some minor changes on top of @abadger PR and did some basic sanity test for `junos_command`, `junos_config` and `junos_facts` modules.

TJuberg commented:
@ganeshrn The additional changes yields the following output
```
2017-05-18 11:16:36,602 p=23406 u=tjuberg |  PLAYBOOK: junos_facts.yml ************************************************************************
2017-05-18 11:16:36,603 p=23406 u=tjuberg |  1 plays in junos_facts.yml
2017-05-18 11:16:36,650 p=23406 u=tjuberg |  PLAY [Get JunOS facts] ***************************************************************************
2017-05-18 11:16:36,726 p=23406 u=tjuberg |  META: ran handlers
2017-05-18 11:16:36,728 p=23406 u=tjuberg |  TASK [Gather Junos facts] ************************************************************************
2017-05-18 11:16:36,728 p=23406 u=tjuberg |  task path: /home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.yml:9
2017-05-18 11:16:37,409 p=23509 u=tjuberg |  creating new control socket for host xxx.xxx.xxx.xxx:22 as user ansible
2017-05-18 11:16:37,409 p=23509 u=tjuberg |  control socket path is /home/tjuberg/.ansible/pc/b8a6122e69
2017-05-18 11:16:37,410 p=23509 u=tjuberg |  current working directory is /home/tjuberg/work/gitlab/bachelor_ansible
2017-05-18 11:16:37,410 p=23509 u=tjuberg |  using connection plugin netconf
2017-05-18 11:16:37,539 p=23509 u=tjuberg |  network_os is set to junos
2017-05-18 11:16:37,539 p=23509 u=tjuberg |  ssh connection done, stating ncclient
2017-05-18 11:16:37,592 ncclient.transport.ssh Connected (version 2.0, client OpenSSH_7.1)
2017-05-18 11:16:38,905 ncclient.transport.ssh Authentication (password) successful!
2017-05-18 11:16:39,815 ncclient.transport.session initialized: session-id=8406 | server_capabilities=<dict_keyiterator object at 0x7ff7adf3fd68>
2017-05-18 11:16:39,816 p=23509 u=tjuberg |  ncclient manager object created successfully
2017-05-18 11:16:39,816 p=23509 u=tjuberg |  connection established to xxx.xxx.xxx.xxx in 0:00:02.406408
2017-05-18 11:16:40,410 p=23509 u=tjuberg |  incoming request accepted on persistent socket
2017-05-18 11:16:40,410 p=23509 u=tjuberg |  socket operation is CONTEXT
2017-05-18 11:16:40,410 p=23509 u=tjuberg |  socket operation is EXEC
2017-05-18 11:16:40,410 p=23509 u=tjuberg |  --- exec_command request open_session()
2017-05-18 11:16:40,411 p=23509 u=tjuberg |  socket operation completed with rc 1
2017-05-18 11:16:40,477 p=23406 u=tjuberg |  open_session() returned 1 b'' b'unable to parse request'
2017-05-18 11:16:40,482 p=23406 u=tjuberg |  fatal: [mx104]: FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell",
    "rc": 1
}
2017-05-18 11:16:40,484 p=23406 u=tjuberg |  	to retry, use: --limit @/home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.retry

2017-05-18 11:16:40,485 p=23406 u=tjuberg |  PLAY RECAP ***************************************************************************************
2017-05-18 11:16:40,485 p=23406 u=tjuberg |  mx104                      : ok=0    changed=0    unreachable=0    failed=1   
```

ganeshrn commented:
@TJuberg Just to be sure did you add this change?
```
-  if request == 'open_session()':
+  if request == b'open_session()':
```
For `open_session()` it should return from [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/netconf.py#L111)

TJuberg commented:
@ganeshrn I can confirm that the code reads
```python
          display.display("--- exec_command request %s" % request, log_only=True)                 
          if request == b'open_session()':                                                        
              return (0, b'ok', b'')   
```

However I managed to miss the change to req. 
Apologies for the confusion.

Updated and more sensible log with linebreaks added for readability
```
2017-05-18 12:17:19,706 p=7243 u=tjuberg |  incoming request accepted on persistent socket
2017-05-18 12:17:19,707 p=7243 u=tjuberg |  socket operation is CONTEXT
2017-05-18 12:17:19,707 p=7243 u=tjuberg |  socket operation is EXEC
2017-05-18 12:17:19,707 p=7243 u=tjuberg |  --- exec_command request open_session()
2017-05-18 12:17:19,708 p=7243 u=tjuberg |  --- exec_command req open_session()
2017-05-18 12:17:19,708 ncclient.operations.rpc Requesting 'ExecuteRpc'
2017-05-18 12:17:19,715 p=7243 u=tjuberg |  socket operation completed with rc 255
2017-05-18 12:17:19,768 p=7194 u=tjuberg |  open_session() returned 255 b'' b'Traceback (most recent call last):\n
 File "/home/tjuberg/.local/badger-ansible/bin/ansible-connection", line 199, in run\n
(rc, stdout, stderr) = self.conn.exec_command(cmd)\n  
File "/home/tjuberg/.local/badger-ansible/lib/ansible/plugins/connection/__init__.py", line 52, in wrapped\n    
return func(self, *args, **kwargs)\n  
File "/home/tjuberg/.local/badger-ansible/lib/ansible/plugins/connection/netconf.py", line 126, in exec_command\n    
reply = self._manager.rpc(req)\n  
File "/home/tjuberg/.local/lib/python3.6/site-packages/ncclient/manager.py", line 172, in wrapper\n    
return self.execute(op_cls, *args, **kwds)\n  
File "/home/tjuberg/.local/lib/python3.6/site-packages/ncclient/manager.py", line 232, in execute\n    raise_mode=self._raise_mode).request(*args, **kwds)\n  
File "/home/tjuberg/.local/lib/python3.6/site-packages/ncclient/operations/third_party/juniper/rpc.py", line 44, in request\n    
return self._request(rpc)\n  
File "/home/tjuberg/.local/lib/python3.6/site-packages/ncclient/operations/rpc.py", line 315, in _request\n    
req = self._wrap(op)\n  
File "/home/tjuberg/.local/lib/python3.6/site-packages/ncclient/operations/rpc.py", line 301, in _wrap\n    
ele.append(subele)\n
TypeError: Argument \'element\' has incorrect type (expected lxml.etree._Element, got NoneType)\n'
2017-05-18 12:17:19,770 p=7194 u=tjuberg |  fatal: [mx104]: FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell",
    "rc": 255
}
```

ganeshrn commented:
@TJuberg Apologies for a delayed response. I missed tracking this PR.

Can you please check on the latest devel if you are still seeing this issue. While doing other refactoring many of the changes in this PR are already merged in `devel` branch.

needs_info

tjuberg commented:
@ganeshrn Unfortunately I am no longer in a position to be able to test this further as our bachelor project is over and I no longer have access to the lab setup with the appropriate equipment.

It should be possible to reproduce this issue if it still exists using the docker file located at https://github.com/TJuberg/NTNU-IDI-Bachelor-2017-29E/tree/master/Vagrant/Arch

Otherwise it would be fair to assume that given the overall work done in relation to python3 in the intervening period that most likely has covered the issue encountered.

ganeshrn commented:
@abadger @TJuberg 
I will close this PR as the issue in no longer seen on devel branch.

Thank you for your contribution.
